// mind-ar-react/dist/examples/ImageTracking.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { Camera, GLTFModel, Plane, Assets, Item } from 'aframe-react-component';
import ImageTracking from '../provider/ImageTracking';
import { Entity, Marker, Scene } from '../components';
import useARManager from '../utils/useARManager';
const ExampleImageTracking = () => {
    const [enabled, setEnabled] = React.useState(false);
    const sceneRef = React.useRef();
    const { startAR, stopAR } = useARManager(sceneRef);
    const onClick = () => {
        if (enabled) {
            stopAR();
        }
        else {
            startAR();
        }
        setEnabled((curr) => !curr);
    };
    const rotationSettings = {
        enabled: true,
        rotationFactor: 5,
    };
    const scaleSettings = {
        enabled: true,
        minScale: 0.3,
        maxScale: 8,
    };
    return (_jsxs(ImageTracking, { children: [_jsx("button", { onClick: onClick, style: { position: 'absolute', zIndex: 999 }, children: enabled ? 'Stop' : 'Start' }), _jsxs(Scene, { mindARImage: {
                    imageTargetSrc: 'https://cdn.jsdelivr.net/gh/hiukim/mind-ar-js@1.0.0/examples/image-tracking/assets/card-example/card.mind',
                    autoStart: false,
                }, "color-space": "sRGB", "mouse-detector": true, "gesture-detector": true, embedded: true, renderer: "colorManagement: true, physicallyCorrectLights", orientationUI: true, stats: enabled, ref: sceneRef, children: [_jsxs(Assets, { children: [_jsx("img", { id: "card", src: "https://cdn.jsdelivr.net/gh/hiukim/mind-ar-js@1.1.4/examples/image-tracking/assets/card-example/card.png", alt: "" }), _jsx(Item, { id: "avatarModel", src: "https://cdn.jsdelivr.net/gh/hiukim/mind-ar-js@1.1.4/examples/image-tracking/assets/card-example/softmind/scene.gltf" })] }), _jsx(Camera, { position: { x: 0, y: 0, z: 0 }, "look-controls": false }), _jsx(Entity, { visible: enabled, children: _jsx(Marker, { targetIndex: 0, children: _jsxs(Entity, { "mouse-rotation": rotationSettings, "mouse-scale": scaleSettings, "gesture-rotation": rotationSettings, "gesture-scale": scaleSettings, children: [_jsx(Plane, { src: "#card", position: [0, 0, 0], height: 0.552, width: 1, rotation: [0, 0, 0] }), _jsx(GLTFModel, { rotation: [0, 0, 0], position: [0, 0, 0.1], scale: [0.005, 0.005, 0.005], animation: {
                                            property: 'position',
                                            to: '0 0.1 0.1',
                                            dur: 1000,
                                            easing: 'easeInOutQuad',
                                            loop: true,
                                            dir: 'alternate',
                                        }, src: "#avatarModel" })] }) }) })] })] }));
};
export default ExampleImageTracking;
